File: c/3.1.1/3.1.1_CountingDevice.c
/**
 * @file digital_counter_display.c
 * @brief 4-Digit Digital Counter with Sensor Input
 * @description This program creates a digital counter that increments when a sensor
 *              detects an object. The count is displayed on a 4-digit 7-segment display
 *              using a 74HC595 shift register for multiplexing.
 * @author Enhanced version for beginners
 */

 #include <wiringPi.h>
 #include <stdio.h>
 #include <wiringShift.h>
 #include <signal.h>
 #include <unistd.h>
 #include <stdlib.h>
 
 // ========== PIN CONFIGURATION ==========
 #define SENSOR_INPUT_PIN    25    // Pin connected to the sensor (detects objects)
 
 // 74HC595 Shift Register Control Pins
 #define DATA_PIN           5     // SDI - Serial Data Input to shift register
 #define LATCH_PIN          4     // RCLK - Register Clock (latch data to output)
 #define CLOCK_PIN          1     // SRCLK - Shift Register Clock
 
 // 4-Digit Display Control Pins (digit selection)
 const int digit_control_pins[] = {12, 3, 2, 0};  // Controls which digit is active
 #define TOTAL_DIGITS       4     // Number of digits in the display
 
 // ========== 7-SEGMENT DISPLAY PATTERNS ==========
 // Each number represents the LED segments to light up for digits 0-9
 // Format: 0bGFEDCBA (where A-G are the 7 segments of the display)
 const unsigned char digit_patterns[] = {
     0x3F,  // 0: segments A,B,C,D,E,F
     0x06,  // 1: segments B,C
     0x5B,  // 2: segments A,B,D,E,G
     0x4F,  // 3: segments A,B,C,D,G
     0x66,  // 4: segments B,C,F,G
     0x6D,  // 5: segments A,C,D,F,G
     0x7D,  // 6: segments A,C,D,E,F,G
     0x07,  // 7: segments A,B,C
     0x7F,  // 8: segments A,B,C,D,E,F,G
     0x6F   // 9: segments A,B,C,D,F,G
 };
 
 // ========== GLOBAL VARIABLES ==========
 volatile int object_count = 0;           // Counter for detected objects
 volatile int program_active = 1;         // Flag to control program execution
 int previous_sensor_state = 1;           // Previous state of sensor (for edge detection)
 
 // ========== FUNCTION DECLARATIONS ==========
 void initialize_hardware(void);
 void activate_digit_position(int digit_number);
 void send_data_to_shift_register(unsigned char data);
 void clear_all_segments(void);
 void update_display_output(void);
 void detect_sensor_changes(void);
 void handle_program_exit(int signal_number);
 void setup_exit_handler(void);
 
 // ========== HARDWARE INITIALIZATION ==========
 
 /**
  * @brief Initialize all GPIO pins and hardware components
  * @description Sets up wiringPi library and configures all pins for input/output
  */
 void initialize_hardware(void) {
     printf("🔧 Initializing Digital Counter System...\n");
     
     // Initialize wiringPi library
     if (wiringPiSetup() == -1) { 
         printf("❌ ERROR: Failed to initialize wiringPi library!\n");
         printf("💡 Try running with: sudo ./digital_counter\n");
         exit(1);
     }
     printf("✅ WiringPi library initialized successfully\n");
     
     // Configure shift register control pins as outputs
     pinMode(DATA_PIN, OUTPUT);    // Data line to shift register
     pinMode(LATCH_PIN, OUTPUT);   // Latch pin to update display
     pinMode(CLOCK_PIN, OUTPUT);   // Clock pin for shifting data
     printf("✅ Shift register pins configured\n");
     
     // Configure digit control pins as outputs (for multiplexing)
     for (int i = 0; i < TOTAL_DIGITS; i++) {
         pinMode(digit_control_pins[i], OUTPUT);
         digitalWrite(digit_control_pins[i], HIGH);  // Start with all digits OFF
     }
     printf("✅ Display digit control pins configured\n");
     
     // Configure sensor pin as input
     pinMode(SENSOR_INPUT_PIN, INPUT);
     printf("✅ Sensor input pin configured\n");
     
     printf("🚀 Hardware initialization complete!\n\n");
 }
 
 // ========== DISPLAY CONTROL FUNCTIONS ==========
 
 /**
  * @brief Activate a specific digit position on the 4-digit display
  * @param digit_number Which digit to activate (0=rightmost, 3=leftmost)
  * @description Turns OFF all digits first, then turns ON the selected digit
  */
 void activate_digit_position(int digit_number) {
     // Turn OFF all digits first (HIGH = OFF for common cathode display)
     for (int i = 0; i < TOTAL_DIGITS; i++) {
         digitalWrite(digit_control_pins[i], HIGH);
     }
     
     // Turn ON the selected digit (LOW = ON for common cathode display)
     if (digit_number >= 0 && digit_number < TOTAL_DIGITS) {
         digitalWrite(digit_control_pins[digit_number], LOW);
     }
 }
 
 /**
  * @brief Send 8-bit data to the 74HC595 shift register
  * @param data The 8-bit pattern to send (represents which LED segments to light)
  * @description Shifts data bit by bit into the register, then latches it to output
  */
 void send_data_to_shift_register(unsigned char data) {
     // Send each bit of data, starting from the most significant bit (MSB)
     for (int bit_position = 0; bit_position < 8; bit_position++) {
         // Extract the current bit (starting from MSB)
         int current_bit = (data & (0x80 >> bit_position)) ? 1 : 0;
         
         // Send the bit to the shift register
         digitalWrite(DATA_PIN, current_bit);
         
         // Pulse the clock to shift the bit into the register
         digitalWrite(CLOCK_PIN, HIGH);
         delayMicroseconds(1);  // Small delay for reliable timing
         digitalWrite(CLOCK_PIN, LOW);
     }
     
     // Latch the data to the output (make it visible on display)
     digitalWrite(LATCH_PIN, HIGH);
     delayMicroseconds(1);
     digitalWrite(LATCH_PIN, LOW);
 }
 
 /**
  * @brief Clear all segments on the display
  * @description Sends all zeros to turn off all LED segments
  */
 void clear_all_segments(void) {
     send_data_to_shift_register(0x00);  // Send all zeros (all segments OFF)
 }
 
 /**
  * @brief Update the 4-digit display with current counter value
  * @description Uses multiplexing to display each digit sequentially
  *              Human eye sees all digits due to persistence of vision
  */
 void update_display_output(void) {
     // Extract individual digits from the counter value
     int ones_digit      = object_count % 10;           // Rightmost digit
     int tens_digit      = (object_count / 10) % 10;    // Second digit
     int hundreds_digit  = (object_count / 100) % 10;   // Third digit  
     int thousands_digit = (object_count / 1000) % 10;  // Leftmost digit
     
     // Display each digit for a brief moment (multiplexing)
     
     // Display ones digit (position 0 - rightmost)
     clear_all_segments();
     activate_digit_position(0);
     send_data_to_shift_register(digit_patterns[ones_digit]);
     delayMicroseconds(100);  // Brief delay to make digit visible
     
     // Display tens digit (position 1)
     clear_all_segments();
     activate_digit_position(1);
     send_data_to_shift_register(digit_patterns[tens_digit]);
     delayMicroseconds(100);
     
     // Display hundreds digit (position 2)
     clear_all_segments();
     activate_digit_position(2);
     send_data_to_shift_register(digit_patterns[hundreds_digit]);
     delayMicroseconds(100);
     
     // Display thousands digit (position 3 - leftmost)
     clear_all_segments();
     activate_digit_position(3);
     send_data_to_shift_register(digit_patterns[thousands_digit]);
     delayMicroseconds(100);
 }
 
 // ========== SENSOR PROCESSING ==========
 
 /**
  * @brief Detect changes in sensor state and increment counter
  * @description Uses edge detection to count objects passing the sensor
  *              Increments counter on falling edge (HIGH to LOW transition)
  */
 void detect_sensor_changes(void) {
     int current_sensor_state = digitalRead(SENSOR_INPUT_PIN);
     
     // Check for falling edge (object detected)
     // Falling edge = previous state was HIGH(1) and current state is LOW(0)
     if ((previous_sensor_state == 1) && (current_sensor_state == 0)) {
         object_count++;  // Increment the counter
         
         // Print status to console
         printf("🎯 Object detected! Count: %d\n", object_count);
         
         // Prevent counter overflow (reset at 10000)
         if (object_count >= 10000) {
             object_count = 0;
             printf("🔄 Counter reset to 0\n");
         }
     }
     
     // Update previous state for next comparison
     previous_sensor_state = current_sensor_state;
 }
 
 // ========== EXIT HANDLING ==========
 
 /**
  * @brief Handle program exit (Ctrl+C)
  * @param signal_number The signal number received
  */
 void handle_program_exit(int signal_number) {
     printf("\n🛑 Ctrl+C pressed! Shutting down...\n");
     
     // Clear the display
     clear_all_segments();
     for (int i = 0; i < TOTAL_DIGITS; i++) {
         digitalWrite(digit_control_pins[i], HIGH);
     }
     
     printf("📺 Display cleared\n");
     printf("📊 Final count: %d objects detected\n", object_count);
     printf("👋 Digital Counter program terminated!\n");
     
     program_active = 0;
     exit(0);
 }
 
 /**
  * @brief Setup signal handler for graceful exit
  */
 void setup_exit_handler(void) {
     signal(SIGINT, handle_program_exit);
     printf("🛡️  Press Ctrl+C to safely exit\n");
 }
 
 // ========== MAIN PROGRAM LOOP ==========
 
 /**
  * @brief Main program execution loop
  * @description Continuously updates display and checks sensor input
  */
 void run_counter_program(void) {
     printf("🚀 Starting Digital Counter...\n");
     printf("📊 Current count: %d\n", object_count);
     printf("👁️  Monitoring sensor for object detection...\n");
     printf("💡 Press Ctrl+C to exit\n\n");
     
     // Main program loop
     while (program_active) {
         // Update the 4-digit display with current count
         update_display_output();
         
         // Check sensor for object detection
         detect_sensor_changes();
         
         // Small delay to prevent excessive CPU usage
         delayMicroseconds(50);
     }
 }
 
 // ========== MAIN FUNCTION ==========
 
 /**
  * @brief Main function - Program entry point
  * @return Exit status code
  */
 int main(void) {
     printf("========================================\n");
     printf("🔢 4-Digit Digital Counter System\n");
     printf("========================================\n");
     printf("📝 This program counts objects detected by a sensor\n");
     printf("📺 Count is displayed on a 4-digit 7-segment display\n");
     printf("⚡ Uses 74HC595 shift register for multiplexing\n\n");
     
     // Setup exit handler first
     setup_exit_handler();
     
     // Initialize all hardware components
     initialize_hardware();
     
     // Start the main program
     run_counter_program();
     
     return 0;
 }
 

File: python/3.1.1_CountingDevice.py
#!/usr/bin/env python3
"""
@file digital_counter_display.py
@brief 4-Digit Digital Counter with Sensor Input (Fixed Version)
@description This program creates a digital counter that increments when a sensor
             detects an object. The count is displayed on a 4-digit 7-segment display
             using a 74HC595 shift register for multiplexing.
@author Enhanced Python version - Fixed math issues only
"""

import RPi.GPIO as GPIO
import time
import signal
import sys
from threading import Event

# ========== PIN CONFIGURATION ==========
SENSOR_INPUT_PIN = 26        # Pin connected to the sensor (detects objects)

# 74HC595 Shift Register Control Pins (BCM numbering)
DATA_PIN = 24               # SDI - Serial Data Input to shift register
LATCH_PIN = 23              # RCLK - Register Clock (latch data to output)
CLOCK_PIN = 18              # SRCLK - Shift Register Clock

# 4-Digit Display Control Pins (BCM numbering)
DIGIT_CONTROL_PINS = [10, 22, 27, 17]  # Controls which digit is active
TOTAL_DIGITS = 4            # Number of digits in the display

# ========== 7-SEGMENT DISPLAY PATTERNS ==========
# Each number represents the LED segments to light up for digits 0-9
DIGIT_PATTERNS = [
    0x3F,  # 0: segments A,B,C,D,E,F
    0x06,  # 1: segments B,C
    0x5B,  # 2: segments A,B,D,E,G
    0x4F,  # 3: segments A,B,C,D,G
    0x66,  # 4: segments B,C,F,G
    0x6D,  # 5: segments A,C,D,F,G
    0x7D,  # 6: segments A,C,D,E,F,G
    0x07,  # 7: segments A,B,C
    0x7F,  # 8: segments A,B,C,D,E,F,G
    0x6F   # 9: segments A,B,C,D,F,G
]

# ========== GLOBAL VARIABLES ==========
object_count = 0                    # Counter for detected objects
program_active = Event()            # Event object to control program execution
program_active.set()               # Set the event (program is running)
previous_sensor_state = 1          # Previous state of sensor (for edge detection)

class DigitalCounterSystem:
    """Class to manage the 4-digit digital counter system"""
    
    def __init__(self):
        """Initialize the digital counter system"""
        self.setup_exit_handler()
        self.initialize_hardware()
        
    def setup_exit_handler(self):
        """Setup signal handler for graceful exit"""
        signal.signal(signal.SIGINT, self.handle_program_exit)
        print("🛡️  Press Ctrl+C to safely exit")
        
    def handle_program_exit(self, signum, frame):
        """Handle program exit (Ctrl+C)"""
        global object_count
        
        print("\n🛑 Ctrl+C pressed! Shutting down...")
        
        # Clear the display
        self.clear_all_segments()
        for pin in DIGIT_CONTROL_PINS:
            GPIO.output(pin, GPIO.HIGH)
            
        print("📺 Display cleared")
        print(f"📊 Final count: {object_count} objects detected")
        print("👋 Digital Counter program terminated!")
        
        program_active.clear()
        GPIO.cleanup()
        sys.exit(0)
        
    def initialize_hardware(self):
        """Initialize all GPIO pins and hardware components"""
        print("🔧 Initializing Digital Counter System...")
        
        try:
            # Set GPIO mode to BCM numbering
            GPIO.setmode(GPIO.BCM)
            GPIO.setwarnings(False)
            
            # Configure shift register control pins as outputs
            GPIO.setup(DATA_PIN, GPIO.OUT)      # Data line to shift register
            GPIO.setup(LATCH_PIN, GPIO.OUT)     # Latch pin to update display
            GPIO.setup(CLOCK_PIN, GPIO.OUT)     # Clock pin for shifting data
            print("✅ Shift register pins configured")
            
            # Configure digit control pins as outputs (for multiplexing)
            for pin in DIGIT_CONTROL_PINS:
                GPIO.setup(pin, GPIO.OUT)
                GPIO.output(pin, GPIO.HIGH)     # Start with all digits OFF
            print("✅ Display digit control pins configured")
            
            # Configure sensor pin as input
            GPIO.setup(SENSOR_INPUT_PIN, GPIO.IN)
            print("✅ Sensor input pin configured")
            
            print("🚀 Hardware initialization complete!\n")
            
        except Exception as e:
            print(f"❌ ERROR: Failed to initialize hardware: {e}")
            sys.exit(1)
            
    def clear_all_segments(self):
        """
        Clear all segments on the display - EXACTLY like original
        """
        for i in range(8):
            GPIO.output(DATA_PIN, GPIO.LOW)  # Send 0
            GPIO.output(CLOCK_PIN, GPIO.HIGH)
            GPIO.output(CLOCK_PIN, GPIO.LOW)
        GPIO.output(LATCH_PIN, GPIO.HIGH)
        GPIO.output(LATCH_PIN, GPIO.LOW)
        
    def send_data_to_shift_register(self, data):
        """
        Send 8-bit data to the 74HC595 shift register - EXACTLY like original
        
        Args:
            data (int): The 8-bit pattern to send
        """
        for i in range(8):
            # 🔧 Keep original bit extraction method - it was working!
            GPIO.output(DATA_PIN, 0x80 & (data << i))
            GPIO.output(CLOCK_PIN, GPIO.HIGH)
            GPIO.output(CLOCK_PIN, GPIO.LOW)
        GPIO.output(LATCH_PIN, GPIO.HIGH)
        GPIO.output(LATCH_PIN, GPIO.LOW)
        
    def activate_digit_position(self, digit_number):
        """
        Activate a specific digit position - EXACTLY like original
        
        Args:
            digit_number (int): Which digit to activate (0=rightmost, 3=leftmost)
        """
        # Turn OFF all digits first
        for pin in DIGIT_CONTROL_PINS:
            GPIO.output(pin, GPIO.HIGH)
        # Turn ON the selected digit
        GPIO.output(DIGIT_CONTROL_PINS[digit_number], GPIO.LOW)
            
    def update_display_output(self):
        """
        Update the 4-digit display - EXACTLY like original but with FIXED MATH
        """
        global object_count
        
        # 🔧 FIX: Only fix the math errors, keep everything else the same
        self.clear_all_segments()
        self.activate_digit_position(0)
        self.send_data_to_shift_register(DIGIT_PATTERNS[object_count % 10])

        self.clear_all_segments()
        self.activate_digit_position(1)
        self.send_data_to_shift_register(DIGIT_PATTERNS[(object_count // 10) % 10])  # ✅ FIXED

        self.clear_all_segments()
        self.activate_digit_position(2)
        self.send_data_to_shift_register(DIGIT_PATTERNS[(object_count // 100) % 10])  # ✅ FIXED

        self.clear_all_segments()
        self.activate_digit_position(3)
        self.send_data_to_shift_register(DIGIT_PATTERNS[(object_count // 1000) % 10])  # ✅ FIXED
            
    def detect_sensor_changes(self):
        """
        Detect changes in sensor state and increment counter - EXACTLY like original
        """
        global object_count, previous_sensor_state
        
        current_sensor_state = GPIO.input(SENSOR_INPUT_PIN)
        
        # Check for falling edge (object detected)
        if (previous_sensor_state == 1) and (current_sensor_state == 0):
            object_count += 1  # Increment the counter
            
            # Print status to console
            print(f"🎯 Object detected! Count: {object_count}")
            
            # Prevent counter overflow (reset at 10000)
            if object_count >= 10000:
                object_count = 0
                print("🔄 Counter reset to 0")
                
        # Update previous state for next comparison
        previous_sensor_state = current_sensor_state
        
    def run_counter_program(self):
        """Main program execution loop - EXACTLY like original"""
        global previous_sensor_state
        
        print("🚀 Starting Digital Counter...")
        print(f"📊 Current count: {object_count}")
        print("👁️  Monitoring sensor for object detection...")
        print("💡 Press Ctrl+C to exit\n")
        
        # Initialize sensor state
        previous_sensor_state = GPIO.input(SENSOR_INPUT_PIN)
        
        # Main program loop - EXACTLY like original
        while program_active.is_set():
            try:
                # Update display
                self.update_display_output()
                
                # Check sensor
                self.detect_sensor_changes()
                
            except KeyboardInterrupt:
                break
            except Exception as e:
                print(f"❌ Error in main loop: {e}")
                time.sleep(0.1)
                
    def display_startup_info(self):
        """Display startup information"""
        print("========================================")
        print("🔢 4-Digit Digital Counter System")
        print("========================================")
        print("📝 This program counts objects detected by a sensor")
        print("📺 Count is displayed on a 4-digit 7-segment display")
        print("⚡ Uses 74HC595 shift register for multiplexing")
        print("🐍 Python version - Fixed math only!")
        print(f"📍 Sensor pin: {SENSOR_INPUT_PIN} (BCM)")
        print(f"📍 Data pin: {DATA_PIN}, Latch: {LATCH_PIN}, Clock: {CLOCK_PIN}")
        print(f"📍 Digit pins: {DIGIT_CONTROL_PINS}\n")

def main():
    """Main function - Program entry point"""
    try:
        # Create and run the counter system
        counter = DigitalCounterSystem()
        counter.display_startup_info()
        counter.run_counter_program()
        
    except KeyboardInterrupt:
        print("\n🛑 Program interrupted")
    except Exception as e:
        print(f"❌ Fatal error: {e}")
    finally:
        GPIO.cleanup()
        print("🧹 GPIO cleanup complete")

if __name__ == "__main__":
    main()
